### ISSUE TITLE
About the reproduce problem

### ISSUE THREAD

**Issue by xhan233** [YUE-FAN/CoSSL#6]
Hi. Thanks for the great work here.

I have some confusion when I try to reproduce your work.Form the training examples,it means we must to pretrain a module for cossl training,what if pre-training is not applied?Can we directly us a Fixmatch pretrain model to resume without running pretrain phase?And where is the  two-phase training explanation in paper?your framework is still a two-stage method?

Hope to reveive your reply.

**Comment by YUE-FAN** [comment]
Hey, thanks for your interest. 

- _Your framework is still a two-stage method?_

Yes and no. 
No, because we always resume the optimizer and the EMA model in the second script, and everything from the first script is still there in the second. Therefore, it is equivalent to end-to-end training with a warm-up. This codebase implements CoSSL in two scripts to avoid repeating the warm-up training every time when designing the method (saving some GPU time). We also verified that implementing CoSSL in a single script as joint training performed very similarly. Yes, because the training scheme changes after the warm-up. So, I think the ''two-stage'' really depends on how you want to see it. In the paper, we term CoSSL as "one-stage" to contrast with methods that freeze the feature after the pretraining and only finetune/re-train the final layer.

- _Where is the two-phase training explanation in the paper?_

In Appendix D, we discussed the effect of the number of warm-up epochs (or pre-training if you prefer). A certain number of warm-up epochs is necessary for the final performance, which is also appreciated as it reduces the training costs since there is no extra computation incurred during the warm-up. And note that we still ensure the same number of training iterations in training the backbone.

- _What if pre-training is not applied? Can we directly use a Fixmatch pretrained model to resume without running the pretrained phase?_

As mentioned above, you just need to warm-up the features. So it is possible to directly resume a Fixmatch pretrained model. But note that the original implementation of FixMatch uses a cosine learning rate schedule. So if you resume that model, you might still want to resume the scheduler as well. In that case, I can't guarantee the optimal number of warm-up epochs is still the same. Therefore, the performance might be different.

**Comment by xhan233** [comment]
I probably understand what you mean, thank you for your patience！！！

**Comment by xhan233** [comment]
I have another question:what should I do when I want to reproduce the result of Table4,I can't find the coresponding imbalanced test data sampling code in your project

**Comment by YUE-FAN** [comment]
Hey, I just uploaded a sloppy script `eval_shifted_test_distribution.py` I used for the shifted evaluation. Please add the paths to your checkpoints in `model_path_list` in the file before running it.

For cifar-10 checkpoints: `python eval_shifted_test_distribution.py --dataset cifar10 --gpu 0`
For cifar-100 checkpoints: `python eval_shifted_test_distribution.py --dataset cifar100 --gpu 0`

**Comment by xhan233** [comment]
> Hey, I just uploaded a sloppy script `eval_shifted_test_distribution.py` I used for the shifted evaluation. Please add the paths to your checkpoints in `model_path_list` in the file before running it.
> 
> For cifar-10 checkpoints: `python eval_shifted_test_distribution.py --dataset cifar10 --gpu 0` For cifar-100 checkpoints: `python eval_shifted_test_distribution.py --dataset cifar100 --gpu 0`

Thanks!
